### PR TITLE
Added rcount/ccount to plot_surface()

### PR DESCRIPTION
**Pull request by WeatherGod**
... providing an alternative to rstride/cstride

Still needs a whats_new.rst entry.

### CODE REVIEW

**Comment by tacaswell** [comment]
Slightly concerned about making the API too complicated.

**Comment by anntzer** [comment]
Perhaps better as a single "max_gridsize" kwarg?

**Comment by WeatherGod** [comment]
Here is my rationale for my API decision. The rest of the `plot_surface()` method is highly dependent upon rstride/cstride. It would be a significant amount of work to refactor that to take something more general, and yield very little gain. So, if rstride/cstride are still going to be the critical pieces of information within the method, then it makes no sense to deprecate them as inputs... so they stay.

Next, I distaste the idea of overloading the current rstride/cstride arguments, which has been suggested by others. The names of the arguments have very specific and clear meaning, and so they would be difficult to conceptually overload. Besides, what would I overload it with? String representation of ints? That would be awkward and tedious.

So, that means that not only are rstride/cstride are going to stay, but that new arguments will be needed for any new specification. Given the orthogonality of rstride/cstride, it makes sense that the new arguments would also be orthogonal (so, have two arguments instead of just one). It also makes it simple for managing precedence if both or neither the stride and the new argument are specified.

So, all that is left is the nature of the new arguments. I like my choice of "maximum count" instead of "exact count" because I can't guarantee that the array could be split into that many equally spaced pieces. Remember, I still have to plug into the cstride/rstride usage internally, so the samples must be equally spaced. 

So, now the user is free to use either a stride or a count, which is conceptually similar to `arange()` vs. `linspace()` (granted, the max count idea doesn't fit perfectly with linspace). Depending on their situation, they are free to choose whichever approach that best suits them.

I am not entirely sold on the need to apply a "classic mode" as fully as I do here. I could be convinced to only apply it if neither stride nor count is given.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -1587,8 +1596,30 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
...
         rows, cols = Z.shape
 
+        has_rstride = 'rstride' in kwargs
+        has_cstride = 'cstride' in kwargs
+        has_rcount = 'rcount' in kwargs
+        has_ccount = 'ccount' in kwargs
+
         rstride = kwargs.pop('rstride', 10)
         cstride = kwargs.pop('cstride', 10)
+        rcount = kwargs.pop('rcount', 50)
```
Comment (tacaswell) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: slight preference to raise if both count and stride are provided.

**Comment by tacaswell** [comment]
@WeatherGod I am convinced and am ok with this going in as-is.  Slight preference for raising if both stride and count are given which would make it a bit easier to document and no need to have the classic compatibility mode.

**Comment by dopplershift** [comment]
I agree that passing both flags is an error which shouldn't pass silently.

**Comment by WeatherGod** [comment]
Agreed. I'll see if I can find time tonight to fix that.

**Comment by QuLogic** [comment]
Needs a rebase now.

**Comment by WeatherGod** [comment]
Just realized something... I should add this support to plot_wireframe(), too, right?

**Comment by WeatherGod** [comment]
Of course, the question then becomes: if we do add it to `plot_wireframe()`, then should rcount/ccount be the default samplers, or should it remain rstride/cstride (which currently defaults to 1, unlike `plot_surface()`)?

**Comment by tacaswell** [comment]
I am :+1: on adding it to `plot_wireframe` if we can get it done in time.  That should not hold this PR.

Failures look real and related to this PR.

**Comment by WeatherGod** [comment]
Yes, I forgot to update the tests when I made it such that you can't
specify both counts and strides. I have that fix in queue. I am just trying
to figure out what is desired for plot_wireframe() because it is similar,
but not the same as plot_surface()'s argument handling.

On Thu, Nov 10, 2016 at 10:03 PM, Thomas A Caswell <notifications@github.com

**Comment by tacaswell** [comment]
bump @WeatherGod

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 2]
```diff
@@ -1553,15 +1553,28 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
...
         sample the input data to generate the graph.  If 1k by 1k
-        arrays are passed in the default values for the strides will
-        result in a 100x100 grid being plotted.
+        arrays are passed in, the default values for the strides will
+        result in a 100x100 grid being plotted. Defaults to 10.
+        Raises a ValueError if both stride and count kwargs
+        (see next section) are provided.
+
+        The `rcount` and `ccount` kwargs supersedes `rstride` and
+        `cstride`for default sampling method for surface plotting.
```
Comment (dopplershift) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Missing space after backtick.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 3]
```diff
@@ -1733,7 +1768,21 @@ def plot_wireframe(self, X, Y, Z, *args, **kwargs):
...
         sample the input data to generate the graph. If either is 0
         the input data in not sampled along this direction producing a
-        3D line plot rather than a wireframe plot.
+        3D line plot rather than a wireframe plot. The stride arguments
+        are only used by default if in the 'classic' mode. They are
+        now superceeded by `rcount` and `ccount`. Will raise ValueError
+        if both stride and count are used.
+
+`       The `rcount` and `ccount` kwargs supersedes `rstride` and
+        `cstride`for default sampling method for wireframe plotting.
```
Comment (WeatherGod) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: hmm, maybe

**Comment by QuLogic** [comment]
The single backticks should be double backticks in many cases; they aren't references, but code.

**Comment by WeatherGod** [comment]
Perhaps, but that'll be more suited for a different PR. Right now, I am just following the current conventions in this file.

**Comment by tacaswell** [comment]
@WeatherGod Why did you regenerate the pdf and svg test images?

**Comment by WeatherGod** [comment]
Uhm, good question... the triaging tool told me they were different, but
now I am wondering if it was accidentally picking up "failed" files from an
earlier test run... Will look into it tonight.

**Comment by WeatherGod** [comment]
yeah, looks like those images were accidentally regenerated. I brought back what was in master for them and re-ran my tests and they passed. So I committed those images back in, and then did a squash commit to eliminate the change out of the history. The tests should be done shortly.

**Pull request merged**